File: python/단어 변환.py
global answer
answer = 100


def checkDiffer(a, b):

    cnt = 0

    for i in range(len(a)):
        if a[i] != b[i]:
            cnt += 1

    return cnt


def getAnswer(cur, target, words, check, count):
    global answer
    if cur == target:
        answer = min(answer, count)
        return

    for idx, w in enumerate(words):
        if checkDiffer(cur, w) == 1:
            if check[idx] == False:
                check[idx] = True
                getAnswer(w, target, words, check, count+1)
                check[idx] = False


def solution(begin, target, words):

    if target not in words:
        return 0

    check = [False for _ in range(len(words))]
    getAnswer(begin, target, words, check, 0)

    global answer

    return answer

File:  cpp/단어 변환.cpp
#include <string>
#include <vector>

using namespace std;
int minval=987654321;
bool visit[55]={false, };
string target2="";
void go(string cur, vector<string>words, int movcnt){
    
    if(cur==target2){
        minval = minval>movcnt ? movcnt : minval;
        return;
    }
    
    for(int i=0; i<words.size(); i++){
        int cnt=0; 
        string tmp=words[i];
        for(int j=0; j< cur.size(); j++){
            
                if(cur[j]!=tmp[j]) cnt++;
        }
        
        if(cnt==1 && !visit[i]){
            visit[i]=true;
            go(tmp, words, movcnt+1);
            visit[i]=false;
        }
    
        
        
    }
    
    
}


int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    bool ishere=false;
    target2=target;
    for(int i=0; i<words.size(); i++){
        if(target==words[i]){
            ishere=true;
        }
    }
    if(!ishere) return 0;
    
    go(begin, words, 0);
    
    answer= minval;
    return answer;
}